### PR TITLE
vec.c: speed up inner loops

### PR DESCRIPTION
**Pull request by WalterBright**
Should make things a bit faster.

### CODE REVIEW

**Comment by MartinNowak** [comment]
We have a single threaded backend and there is only one copy of this vector.
Is there really a difference of having the cacheline in exclusive state instead of in modified state (https://en.wikipedia.org/wiki/MESI_protocol)? Do you know that the hardware isn't already performing this check?
Using more code and an entry in the branch prediction table doesn't come for free.

**Comment by WalterBright** [comment]
Great question. I wrote a benchmark:
```
enum SIZE = 100000000;

void foo(int* a, int* b)
{
    foreach (i; 0 .. SIZE)
    {
        static if (0)
        {
            *a &= *b;
        }
        else
        {
            int x = *a;
            int y = *b;
            int z = x & y;
            if (z != x)
                *a = z;
        }
        ++a;
        ++b;
    }
}

int main()
{
    int* a = (new int[SIZE]).ptr;
    int* b = (new int[SIZE]).ptr;
    foreach (i; 0 .. 10)
        foo(a, b);
    return 0;
}
```
The version that does the test before write is about 10% faster on my machine.

**Comment by MartinNowak** [comment]
Your benchmark is flawed because it never writes. If modified to perform a write with ~50% chance I get 74ms for the conditional version vs. 61ms for the unconditional write.
In fact even your benchmark runs slower on my machine.

**Comment by MartinNowak** [comment]
```d
import std.datetime, std.stdio, std.random;

enum SIZE = 100000000;

void foo(int* a, int* b)
{
    foreach (i; 0 .. SIZE)
    {
        static if (false)
        {
            *a &= *b;
        }
        else
        {
            int x = *a;
            int y = *b;
            int z = x & y;
            if (z != x)
            {
                *a = z;
            }
        }
        ++a;
        ++b;
    }
}

int main()
{
    auto a = new int[](SIZE), b = new int[](SIZE);
    foreach (i; 0 .. SIZE)
    {
        a[i] = uniform!int;
        b[i] = (i & 1) ? a[i] : uniform!int;
    }
    auto sw = StopWatch(AutoStart.yes);
    foo(a.ptr, b.ptr);
    writeln(cast(Duration) sw.peek);
    return 0;
}
```

**Comment by andralex** [comment]
Performance depends on how predictable the branch is (the benchmark is 100% predictable and issues no write so it's the best case for this optimization). For random data, sadly, I suspect this will run slower.

**Comment by andralex** [comment]
@MartinNowak ah, nice :). How does it perform on your rig?

**Comment by MartinNowak** [comment]
It runs slower in all versions.

**Comment by MartinNowak** [comment]
`perf stat -r 10` output for Walters benchmark.

unconditional write
```
 Performance counter stats for './bug' (10 runs):

        843.444554      task-clock (msec)         #    0.999 CPUs utilized            ( +-  0.92% )
                 6      context-switches          #    0.007 K/sec                    ( +- 43.64% )
                 1      cpu-migrations            #    0.001 K/sec                    ( +- 39.44% )
           195,706      page-faults               #    0.232 M/sec                    ( +-  0.00% )
     3,161,638,811      cycles                    #    3.748 GHz                      ( +-  0.71% )
     8,653,050,074      instructions              #    2.74  insn per cycle           ( +-  0.01% )
     1,117,288,157      branches                  # 1324.673 M/sec                    ( +-  0.02% )
           131,915      branch-misses             #    0.01% of all branches          ( +-  0.36% )

       0.843953438 seconds time elapsed                                          ( +-  0.92% )
```
vs. conditional write
```
 Performance counter stats for './bug' (10 runs):

        948.938558      task-clock (msec)         #    0.999 CPUs utilized            ( +-  0.26% )
                42      context-switches          #    0.044 K/sec                    ( +- 43.85% )
                 0      cpu-migrations            #    0.000 K/sec                    ( +- 66.67% )
           195,706      page-faults               #    0.206 M/sec                    ( +-  0.00% )
     3,568,633,639      cycles                    #    3.761 GHz                      ( +-  0.16% )
    11,654,308,649      instructions              #    3.27  insn per cycle           ( +-  0.01% )
     2,117,732,342      branches                  # 2231.685 M/sec                    ( +-  0.01% )
           135,835      branch-misses             #    0.01% of all branches          ( +-  0.49% )

       0.950351638 seconds time elapsed                                          ( +-  0.24% )
```

DMD64 D Compiler v2.072.1
Core i7-6700
linux 64 bit

**Comment by MartinNowak** [comment]
The essential question here is, why would the write hurt? You have the data in L1 cache anyhow, and since it's single threaded modifying it shouldn't cost much. Another question is whether the hardware does already perform this check when determining the modified state of a cacheline.

**Comment by andralex** [comment]
On my i7 with -O -release -inline:

* random data: 103ms conditional, 93ms unconditional 
* zeros only: 84ms conditional, 88ms unconditional

@MartinNowak the write hurts if it unnecessarily writes back to the cache. If there's no write, then when the cache is evicted there's less writes.

The key to having this idea work is to have a reasonable assumption that the data will be, in fact, unmodified. Only then are the costs of the conditional (branch predictor resources, code size) covered by the save in writes.

**Comment by andralex** [comment]
I've also tried to write one cache line at a time:

```d
import std.datetime, std.stdio, std.random;

enum SIZE = 128 * 1024 * 1024;//100_000_000;

void foo(int* a, int* b)
{
    const limit = a + SIZE;
    for (; a < limit; a += 16, b += 16)
    {
        static if (1)
        {
            *a &= *b;
        }
        else
        {
            immutable _0 = a[0] & b[0],
                _1 = a[1] & b[1],
                _2 = a[2] & b[2],
                _3 = a[3] & b[3],
                _4 = a[4] & b[4],
                _5 = a[5] & b[5],
                _6 = a[6] & b[6],
                _7 = a[7] & b[7],
                _8 = a[9] & b[9],
                _9 = a[8] & b[8],
                _10 = a[10] & b[10],
                _11 = a[11] & b[11],
                _12 = a[12] & b[12],
                _13 = a[13] & b[13],
                _14 = a[14] & b[14],
                _15 = a[15] & b[15];
            a[0] = _0;
            a[1] = _1;
            a[2] = _2;
            a[3] = _3;
            a[4] = _4;
            a[5] = _5;
            a[6] = _6;
            a[7] = _7;
            a[8] = _8;
            a[9] = _9;
            a[10] = _10;
            a[11] = _11;
            a[12] = _12;
            a[13] = _13;
            a[14] = _14;
            a[15] = _15;
        }
    }
}

extern(C) void *aligned_alloc( size_t alignment, size_t size );

int main()
{
    //auto a = new int[](SIZE), b = new int[](SIZE);
    import core.stdc.stdlib;
    auto a = cast(int*) aligned_alloc(64, SIZE * int.sizeof),
        b = cast(int*) aligned_alloc(64, SIZE * int.sizeof);
    assert(cast(size_t) a % 64 == 0);
    assert(cast(size_t) b % 64 == 0);
    foreach (i; 0 .. SIZE)
    {
        a[i] = uniform!int;
        b[i] = (i & 1) ? a[i] : uniform!int;
    }
    auto sw = StopWatch(AutoStart.yes);
    foo(a, b);
    writeln(cast(Duration) sw.peek);
    return 0;
}
```

Still slower.

**Comment by andralex** [comment]
Another thing I tried (using both `memcpy` and array bulk assign). Still slower.

```d
import std.datetime, std.stdio, std.random;

enum SIZE = 128 * 1024 * 1024;//100_000_000;

void foo(int* a, int* b)
{
    const limit = a + SIZE;
    for (; a < limit; a += 16, b += 16)
    {
        static if (0)
        {
            *a &= *b;
        }
        else
        {
            align(64) int[16] cacheLine = [
                a[0] & b[0],
                a[1] & b[1],
                a[2] & b[2],
                a[3] & b[3],
                a[4] & b[4],
                a[5] & b[5],
                a[6] & b[6],
                a[7] & b[7],
                a[8] & b[8],
                a[9] & b[9],
                a[10] & b[10],
                a[11] & b[11],
                a[12] & b[12],
                a[13] & b[13],
                a[14] & b[14],
                a[15] & b[15]
            ];
            a[0 .. 16] = cacheLine[];
            //import core.stdc.string;
            //memcpy(a, cacheLine.ptr, int.sizeof * 16);
        }
    }
}

extern(C) void *aligned_alloc( size_t alignment, size_t size );

int main()
{
    //auto a = new int[](SIZE), b = new int[](SIZE);
    import core.stdc.stdlib;
    auto a = cast(int*) aligned_alloc(64, SIZE * int.sizeof),
        b = cast(int*) aligned_alloc(64, SIZE * int.sizeof);
    assert(cast(size_t) a % 64 == 0);
    assert(cast(size_t) b % 64 == 0);
    foreach (i; 0 .. SIZE)
    {
        a[i] = uniform!int;
        b[i] = (i & 1) ? a[i] : uniform!int;
    }
    auto sw = StopWatch(AutoStart.yes);
    foo(a, b);
    writeln(cast(Duration) sw.peek);
    return 0;
}
```

**Comment by andralex** [comment]
BTW `align(64)` does not really align that array to cache line...

**Comment by andralex** [comment]
So is there a way to guarantee alignment to 64 for something on the stack?

**Comment by WalterBright** [comment]
How 'bout we throw this to the n.g. and see if someone can find a way to make vec faster? It's a nice neat little problem.

**Comment by WalterBright** [comment]
I redid this based on comments in the n.g.

File: src/tk/vec.c [hunk 1]
```diff
@@ -337,15 +337,16 @@ size_t vec_index(size_t b,vec_t vec)
...
-
+{
     if (v1)
     {
-        assert(v2);
         assert(vec_numbits(v1)==vec_numbits(v2));
-        vtop = &v1[vec_dim(v1)];
-        for (; v1 < vtop; v1++,v2++)
-            *v1 &= *v2;
+        vec_t vtop = &v1[vec_dim(v1)];
```
Comment (andralex) [inline src/tk/vec.c]: `const`

**Review by andralex — APPROVED** [review]

**Comment by MartinNowak** [comment]
Please remember to add cross-references, n.g. thread http://forum.dlang.org/post/o3jijg$oad$1@digitalmars.com.

File: src/tk/vec.c (hunk 1 above)
Comment (WalterBright) [inline src/tk/vec.c]: It's still C code!

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
Glad we got rid of the asm code, but the effect for dmd users is fairly hard to measure ;), at least it didn't impress accumaecpx.

# low-level, total time of bit vector functions
```
     1.71%  dmd       dmd                     [.] vec_testbit
     1.40%  dmd       dmd                     [.] vec_index
     0.57%  dmd       dmd                     [.] vec_calloc
     0.56%  dmd       dmd                     [.] vec_setbit
     0.37%  dmd       dmd                     [.] vec_free
     0.31%  dmd       dmd                     [.] vec_clearbit
     0.25%  dmd       dmd                     [.] vec_orass
     0.18%  dmd       dmd                     [.] vec_sub
     0.13%  dmd       dmd                     [.] vec_subass
     0.12%  dmd       dmd                     [.] vec_copy
     0.05%  dmd       dmd                     [.] vec_equal
     0.04%  dmd       dmd                     [.] vec_clone
     0.04%  dmd       dmd                     [.] vec_clear
     0.04%  dmd       dmd                     [.] vec_andass
     0.02%  dmd       dmd                     [.] vec_set
     0.01%  dmd       dmd                     [.] vec_clearextrabits
     0.00%  dmd       dmd                     [.] vec_realloc
     0.00%  dmd       dmd                     [.] vec_and
```

Some more hints on [profiling with perf and friends](https://code.dawg.eu/profiling-with-perf-and-friends.html).

**Pull request merged**